abaco.go

```diff
@@ -15,59 +16,290 @@ import (
 	"github.com/usnistgov/dastard/ringbuffer"
 )
 
-// ChannelGroup represents a range of consecutive channel numbers
-type ChannelGroup struct {
-	firstchan int  // first channel number in this group
-	nchan int // how many channels in this group
+//------------------------------------------------------------------------------------------------
+
+// GroupIndex represents the specifics of a channel group.
+// It should be globally unique across all Abaco data.
+type GroupIndex struct {
+	firstchan int // first channel number in this group
+	nchan     int // how many channels in this group
+}
+
+// gindex converts a packet to the GroupIndex whose data it contains
+func gIndex(p *packets.Packet) GroupIndex {
+	nchan, offset := p.ChannelInfo()
+	return GroupIndex{nchan: nchan, firstchan: offset}
 }
 
-// ByGroup implements sort.Interface for ChannelGroup slices
-type ByGroup []ChannelGroup
-func (g ByGroup) Len() int { return len(g) }
-func (g ByGroup) Swap(i, j int) { g[i], g[j] = g[j], g[i] }
+// ByGroup implements sort.Interface for []GroupIndex so we can sort such slices.
+type ByGroup []GroupIndex
+
+func (g ByGroup) Len() int           { return len(g) }
+func (g ByGroup) Swap(i, j int)      { g[i], g[j] = g[j], g[i] }
 func (g ByGroup) Less(i, j int) bool { return g[i].firstchan < g[j].firstchan }
```

Comment:
What if one of them has `firstchan = 1` and `nchan=10000000` and the other has `firstchan = 2`. Then one is not strictly less than the other.

Reply:
Right. This is _an_ ordering rule, not _the only conceivable_ ordering rule.

Later, AbacoGroup objects will be sorted in this way, but also checked to ensure that no channel appears in more than one group. Given that check, the rule here is actually the only sensible one.